### PR TITLE
Morphological tagger

### PR DESCRIPTION
**Pull request by AlexeySorokin**
Implements a neural morphological tagger.

### CODE REVIEW

**Comment by seliverstov** [comment]
Add link to you *.md from main README.md (add new row to table with components in 'Features' section)

**Comment by yoptar** [comment]
Why are most configs in the `morpho_tagger/UD2.0/`, but one config is located in `morpho_tagger/`?

File: deeppavlov/dataset_readers/morphotagging_dataset_reader.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                data_path = [data_path / "{}-ud-{}.conllu".format(language, mode)
+                             for mode in data_types]
+                reserve_data_path = [
+                    reserve_data_path / language / "{}-ud-{}.conllu".format(language, mode)
+                    for mode in data_types]
+        else:
+            data_path = [Path(data_path) for data_path in data_path]
+            reserve_data_path = None
+        if len(data_path) != len(data_types):
+            print(str(data_path))
```
Comment (mu-arkhipov) [inline deeppavlov/dataset_readers/morphotagging_dataset_reader.py]: Please use logger instead of print
`log = get_logger(__name__)`

File: deeppavlov/models/morpho_tagger/network.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        self.highway_dropout = highway_dropout
+        self.intermediate_dropout = intermediate_dropout
+        self.word_lstm_layers = word_lstm_layers
+        self.word_lstm_units = word_lstm_units
+        self.lstm_dropout = lstm_dropout
+        self.word_dropout = word_dropout
+        self.regularizer = regularizer
+        self.verbose = verbose
+        self.initialize()
+        print("{} symbols, {} tags".format(self.symbols_number_, self.tags_number_))
```
Comment (mu-arkhipov) [inline deeppavlov/models/morpho_tagger/network.py]: print -> logger

File: deeppavlov/models/morpho_tagger/network.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        word_inputs = kl.Input(shape=(None, MAX_WORD_LENGTH+2), dtype="int32")
+        inputs = [word_inputs]
+        word_outputs = self.build_word_cnn(word_inputs)
+        outputs, lstm_outputs = self.build_basic_network(word_outputs)
+        compile_args = {"optimizer": ko.nadam(lr=0.002, clipnorm=5.0),
+                        "loss": "categorical_crossentropy", "metrics": ["accuracy"]}
+        self.model_ = Model(inputs, outputs)
+        self.model_.compile(**compile_args)
+        if self.verbose > 0:
+            print(self.model_.summary())
```
Comment (mu-arkhipov) [inline deeppavlov/models/morpho_tagger/network.py]: print -> logger

File: deeppavlov/models/morpho_tagger/tagger.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        """Check existence of the model file, load the model if the file exists"""
+
+        # General way (load path from config assumed to be the path
+        # to the file including extension of the file model)
+        model_file_exist = self.load_path.exists()
+        path = str(self.load_path.resolve())
+        print(model_file_exist, path)
+        # Check presence of the model files
+        if model_file_exist:
+            print('[loading model from {}]'.format(path), file=sys.stderr)
```
Comment (mu-arkhipov) [inline deeppavlov/models/morpho_tagger/tagger.py]: print -> logger

File: deeppavlov/configs/morpho_tagger/UD2.0/ru_syntagrus/train_config.json [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    "batch_size": 32,
+
+    "metrics": ["per_token_accuracy"],
+    "validation_patience": 10,
+    "val_every_n_epochs": 1,
+    "log_every_n_epochs": 1
+  },
+  "metadata": {
+    "download": [
+      "http://lnsigo.mipt.ru/export/deeppavlov_data/morpho_tagger.tar.gz"
```
Comment (mu-arkhipov) [inline deeppavlov/configs/morpho_tagger/UD2.0/ru_syntagrus/train_config.json]: Model downloaded from this config always predicts PAD:
x::на той стороне реки
\>\>['PAD', 'PAD', 'PAD', 'PAD']

Comment (mu-arkhipov) [reply deeppavlov/configs/morpho_tagger/UD2.0/ru_syntagrus/train_config.json]: After training predictions seem to be ok:
x::на той стороне реки
\>\> ['ADP', 'DET,Case=Loc|Gender=Fem|Number=Sing', 'NOUN,Animacy=Inan|Case=Loc|Gender=Fem|Number=Sing', 'NOUN,Animacy=Inan|Case=Gen|Gender=Fem|Number=Sing']

Comment (AlexeySorokin) [reply deeppavlov/configs/morpho_tagger/UD2.0/ru_syntagrus/train_config.json]: It seems that weights were not downloaded to the model, I will have a look tomorrow.

File: deeppavlov/configs/morpho_tagger/UD2.0/hu/test_config.json [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      }
+    ],
+    "out": ["y_predicted"]
+  },
+  "train": {
+    "test_best": true,
+    "batch_size": 16,
+    "metrics": ["per_token_accuracy"]
+  }
+}
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/configs/morpho_tagger/UD2.0/hu/test_config.json]: Yes, it should be, I will add the required subsection.

File: deeppavlov/models/preprocessors/capitalization.py [hunk 7]
```diff
@@ -60,3 +63,42 @@ def __call__(self, tokens_batch, **kwargs):
...
+def process_word(word, to_lower=False, append_case=None):
+    if all(x.isupper() for x in word) and len(word) > 1:
+        uppercase = "<ALL_UPPER>"
+    elif word[0].isupper():
+        uppercase = "<FIRST_UPPER>"
+    else:
+        uppercase = None
+    if to_lower:
+        word = word.lower()
+    if word.isdigit():
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/models/preprocessors/capitalization.py]: It is hard to determine what other classes behave just as digits, because 13.25 may be also time which has different morphological contexts compared to usual numbers. I decide not to complicate this too much.

File: deeppavlov/core/commands/train.py [hunk 8]
```diff
@@ -150,17 +152,28 @@ def train_model_from_config(config_path: str) -> None:
...
 
-    if callable(getattr(model, 'train_on_batch', None)):
-        _train_batches(model, iterator, train_config, metrics_functions)
-    elif callable(getattr(model, 'fit_batches', None)):
-        _fit_batches(model, iterator, train_config)
-    elif callable(getattr(model, 'fit', None)):
-        _fit(model, iterator, train_config)
-    elif not isinstance(model, Chainer):
-        log.warning('Nothing to train')
+    if not is_trained:
```
Comment (yoptar) [inline deeppavlov/core/commands/train.py]: Train if not is_trained?

File: deeppavlov/core/commands/train.py [hunk 9]
```diff
@@ -40,6 +40,20 @@
...
+def prettify_metrics(metrics, precision=4):
+    """
+    Prettifies the dictionary of metrics
+    """
+    prettified_metrics = OrderedDict()
+    for key, value in metrics:
+        if key.endswith("accuracy"):
+            value = round(100*value, precision)
+        else:
+            value = round(value, precision)
```
Comment (yoptar) [inline deeppavlov/core/commands/train.py]: Rounding is fine, but forcing to multiply accuracy by 100 is questionable. `0.9` and `0.009` are very different.

File: deeppavlov/core/commands/train.py (hunk 8 above)
Comment (AlexeySorokin) [reply deeppavlov/core/commands/train.py]: Yes, if the model is not trained we should train it, otherwise we use it for evaluation.

File: deeppavlov/core/commands/train.py [hunk 10]
```diff
@@ -204,7 +217,7 @@ def _test_model(model: Component, metrics_functions: List[Tuple[str, Callable]],
 
     report = {
         'eval_examples_count': len(val_y_true),
-        'metrics': OrderedDict(metrics),
+        'metrics': OrderedDict(prettify_metrics(metrics)),
```
Comment (yoptar) [inline deeppavlov/core/commands/train.py]: prettify_metrics already returns `OrderedDict`

File: deeppavlov/core/commands/train.py (hunk 9 above)
Comment (AlexeySorokin) [reply deeppavlov/core/commands/train.py]: I prefer to see accuracy in percents, it is more readable. How should I fix it.

File: deeppavlov/core/commands/train.py (hunk 10 above)
Comment (AlexeySorokin) [reply deeppavlov/core/commands/train.py]: I will fix it.

File: deeppavlov/models/preprocessors/capitalization.py [hunk 11]
```diff
@@ -60,3 +63,42 @@ def __call__(self, tokens_batch, **kwargs):
...
+    if to_lower and uppercase is not None:
+        if append_case == "first":
+            answer = [uppercase] + answer
+        elif append_case == "last":
+            answer = answer + [uppercase]
+    return tuple(answer)
+
+
+@register('lowercase_preprocessor')
+class LowercasePreprocessor(Component):
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/models/preprocessors/capitalization.py]: I do not understand your comment: there is no classes in str_lower that do what I need. Or you suggest to move my code there?

File: deeppavlov/core/commands/train.py [hunk 12]
```diff
@@ -127,21 +141,9 @@ def train_model_from_config(config_path: str) -> None:
...
-            vocabs[vocab_param_name] = _fit(v, iterator, None)
-
-        model_config = config['model']
-        model = from_params(model_config, vocabs=vocabs, mode='train')
-
     train_config = {
         'metrics': ['accuracy'],
-
-        'validate_best': True,
+        'validate_best': not (is_trained),
```
Comment (yoptar) [inline deeppavlov/core/commands/train.py]: I would leave it `True` by default, because it checks if the model saves and loads correctly.

File: deeppavlov/core/commands/train.py (hunk 8 above)
Comment (yoptar) [reply deeppavlov/core/commands/train.py]: We could want to retrain or train trained model on another dataset. How about `evaluate_mode`?

File: README.md [hunk 13]
```diff
@@ -4,7 +4,10 @@
 **We are in a really early Alpha release. You should be ready for hard adventures. 
 If you have updated to version 0.0.2 or greater - please re-download all pre-trained models**
 
-DeepPavlov is an open-source conversational AI library built on [TensorFlow](https://www.tensorflow.org/) and [Keras](https://keras.io/). It is designed for
+### *We are in a really early Alpha release. You should be ready for hard adventures.*
+### *If you have updated to version 0.0.2 or greater - please re-download all pre-trained models*
+
+DeepPavlov is an open-source conversational AI library built on TensorFlow and Keras. It is designed for
```
Comment (yoptar) [inline README.md]: alpha warning is there twice, TensorFlow and Keras links are lost

File: deeppavlov/core/commands/train.py (hunk 9 above)
Comment (AlexeySorokin) [reply deeppavlov/core/commands/train.py]: It is not my known accuracy, it is normal accuracy shown in a proper way. I strongly suggest to demonstrate accuracy always in percents.

File: deeppavlov/models/morpho_tagger/README.md [hunk 14]
```diff
@@ -0,0 +1,19 @@
+It is an implementation of neural morphological tagger from
```
Comment (yoptar) [inline deeppavlov/models/morpho_tagger/README.md]: This readme is very different from the others. You should at least add a header and badges.

File: deeppavlov/models/morpho_tagger/README.md [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+[Heigold et al., 2017. An extensive empirical evaluation of character-based morphological tagging for 14 languages](http://www.aclweb.org/anthology/E17-1048).
+We distribute the model trained on ru_syntagrus corpus of [Universal Dependencies project](www.universaldependencies.org).
+If you want to use it from scratch, do the following:
+
+1. Download the archive from 
+<http://lnsigo.mipt.ru/export/deeppavlov_data/morpho_tagger.tar.gz> to
+ [../../../download](../../../download).
```
Comment (yoptar) [inline deeppavlov/models/morpho_tagger/README.md]: There is no need to download anything by hands, there are commands for that

Comment (yoptar) [reply deeppavlov/models/morpho_tagger/README.md]: And this link gives 404

File: deeppavlov/models/morpho_tagger/README.md [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+2. To apply a pre-trained ru_syntagrus model to ru_syntagrus test data, run
+``python models/morpho_tagger/main.py -p configs/UD2.0/morpho_tagger/ru_syntagrus/morpho_ru_syntagrus_predict.json``.
+3. To test ru_syntagrus model on ru_syntagrus test subset, run
+``python models/morpho_tagger/main.py -T configs/UD2.0/morpho_tagger/ru_syntagrus/morpho_ru_syntagrus_test.json``.
+4. To retrain model on ru_syntagrus dataset, run
+``python models/morpho_tagger/main.py configs/UD2.0/morpho_tagger/ru_syntagrus/morpho_ru_syntagrus_train.json``.
+Be careful, one epoch takes 8-60 minutes depending on your GPU.
+5. To tag Russian sentences from stdin, run
+``python deep.py configs/morpho_tagger/UD2.0/morpho_tagger/ru_syntagrus/morpho_ru_syntagrus_interact.json``.
```
Comment (yoptar) [inline deeppavlov/models/morpho_tagger/README.md]: From what directory are users expected to run these commands?

File: README.md (hunk 13 above)
Comment (AlexeySorokin) [reply README.md]: I will fix it

File: deeppavlov/models/morpho_tagger/main.py [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+            mode = "test"
+        elif opt in ["-p"]:
+            mode = "predict"
+        else:
+            raise ValueError("Unknown option {}".format(opt))
+    config_path = args[0]
+    if mode == "predict":
+        predict_with_model(config_path)
+    elif mode == "test":
+        train_evaluate_model_from_config(config_path, is_trained=True)
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/main.py]: You suggest to add `evaluate` mode to deep.py?

[the remark this answers is not in the document]
Comment (yoptar) [reply deeppavlov/models/morpho_tagger/main.py]: yes

File: deeppavlov/models/morpho_tagger/README.md (hunk 15 above)
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/README.md]: I will fix it.

File: deeppavlov/models/morpho_tagger/README.md (hunk 16 above)
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/README.md]: I will fix it.

File: deeppavlov/models/morpho_tagger/README.md (hunk 15 above)
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/README.md]: You mean the link to download subdirectory?

Comment (yoptar) [reply deeppavlov/models/morpho_tagger/README.md]: yes. If you open this file in github

File: deeppavlov/models/morpho_tagger/main.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (yoptar) [reply deeppavlov/models/morpho_tagger/main.py]: How about test mode?

[the remark this answers is not in the document]
Comment (yoptar) [reply deeppavlov/models/morpho_tagger/main.py]: Move test mode to deep.py and call it `evaluate`. And then leave in main.py only conllu predict and call the script accordingly

File: deeppavlov/core/commands/train.py [hunk 18]
```diff
@@ -154,7 +155,7 @@ def train_evaluate_model_from_config(config_path: str, is_trained=False) -> None
 
     metrics_functions = list(zip(train_config['metrics'], get_metrics_by_names(train_config['metrics'])))
 
-    if not is_trained:
+    if train_config['train_model']:
```
Comment (yoptar) [inline deeppavlov/core/commands/train.py]: It should not be necessary to change config file to evaluate the model without training

Comment (AlexeySorokin) [reply deeppavlov/core/commands/train.py]: Now `train`, `validate` and 'test' flags are used analogously, I see no reason to set one of them in code and another in config. Either we have three optional parameters of the function called `train_model`, `evaluate_model` and `test_model`, or we have three fields defined in config. Probably,  both modes can be allowed.

Comment (AlexeySorokin) [reply deeppavlov/core/commands/train.py]: By the way, now the code is more concise than it was.

File: deeppavlov/core/commands/train.py (hunk 9 above)
Comment (AlexeySorokin) [reply deeppavlov/core/commands/train.py]: There should be some parameter regulating metrics precision and demonstration, I will think on it.

Comment (seliverstov) [reply deeppavlov/core/commands/train.py]: I think it's bad practice to implicitly modify any metrics that ends with 'accuracy'. Multiplication by 100 should be removed.

File: README.md [hunk 19]
```diff
@@ -4,7 +4,7 @@
 **We are in a really early Alpha release. You should be ready for hard adventures. 
 If you have updated to version 0.0.2 or greater - please re-download all pre-trained models**
 
-DeepPavlov is an open-source conversational AI library built on [TensorFlow](https://www.tensorflow.org/) and [Keras](https://keras.io/). It is designed for
+DeepPavlov is built on top of machine learning frameworks [TensorFlow](https://www.tensorflow.org/) and [Keras](https://keras.io/). Other external libraries can be used to build basic components.
```
Comment (yoptar) [inline README.md]: But what is DeepPavlov?

File: deeppavlov/models/morpho_tagger/README.md [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+It is an implementation of neural morphological tagger from
+[Heigold et al., 2017. An extensive empirical evaluation of character-based morphological tagging for 14 languages](http://www.aclweb.org/anthology/E17-1048).
+We distribute the model trained on ru_syntagrus corpus of [Universal Dependencies project](www.universaldependencies.org).
+If you want to use it from scratch, do the following:
+
+1. Download data
+``python -m deeppavlov.deep download deeppavlov/configs/morpho_tagger/UD2.0/ru_syntagrus/morpho_ru_syntagrus_train.json``
+To perform all downloads in runtime you can also run all subsequent commands with ``-d`` key, 
+2. To apply a pre-trained ru_syntagrus model to ru_syntagrus test data, run
+``python models/morpho_tagger/main.py train configs/morpho_tagger/UD2.0/ru_syntagrus/morpho_ru_syntagrus_predict.json``.
```
Comment (yoptar) [inline deeppavlov/models/morpho_tagger/README.md]: Use \`\`\`code\`\`\` for bigger commands and \`-d\` for smaller blocks

File: deeppavlov/models/morpho_tagger/main.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/main.py]: All except predict has been moved to ``deep.py``, cancel the change request, please.
